### PR TITLE
[ZEPPELIN-2582][DOCS] docs for interpreter binding modes

### PR DESCRIPTION
**Pull request by 1ambda**
### What is this PR for?

Updated `interpreter_binding_mode.md` since users are sometimes confused what this mode means and there is already opened JIRA issue. This documentation will be helpful to Zeppelin users.

**disclaimer: content was copied from [here](https://medium.com/@leemoonsoo/apache-zeppelin-interpreter-mode-explained-bae0525d0555) with author's consent.**

### What type of PR is it?
[Documentation]

### Todos

DONE

### What is the Jira issue?

[ZEPPELIN-2582](https://issues.apache.org/jira/browse/ZEPPELIN-2582)

### How should this be tested?

1. setup rvm 2.1.0+ and install required bundles for `docs/`
2. bundle exec jekyll serve --watch
3. http://localhost:4000/usage/interpreter/interpreter_binding_mode.html

### Screenshots (if appropriate)

image

image

image

image

image

image

### Questions:
* Does the licenses files need update? - NO
* Is there breaking changes for older versions? - NO
* Does this needs documentation? - This PR is about docs.

### CODE REVIEW

**Comment by 1ambda** [comment]
CI failed but irrelevant.

- https://travis-ci.org/1ambda/zeppelin/jobs/246983015

**Comment by cacti77** [comment]
Thanks for doing this. I looked at the screenshots above. The diagrams are very helpful!  Are you also going to add the table I created at https://issues.apache.org/jira/browse/ZEPPELIN-2582?focusedCommentId=16024796&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16024796 or do you think it doesn't add anything useful?

**Comment by 1ambda** [comment]
@cacti77 Yes, it would be a good summary. Let me add it.

**Comment by 1ambda** [comment]
I added your summary table.

image

**Comment by 1ambda** [comment]
CI failed but irrelevant.

**Comment by zjffdu** [comment]
It is better to mention in what dimension of these modes. (per user or per note).
And from the code perspective, there's only one `InterpreterGroup` for scoped mode, but multiple sessions in one InterpreterGroup.

**Comment by 1ambda** [comment]
@zjffdu Thanks. Let me add the description into the doc.

**Comment by 1ambda** [comment]
@zjffdu enhanced description according to your suggestion. could you check that?

image

image

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 1]
```diff
@@ -25,4 +25,88 @@ limitations under the License.
...
+In **Shared** mode, single JVM process and single Interpreter Group serves all Notes.
+
+## Scoped Mode
+
+<div class="text-center">
+    <img src="/assets/themes/zeppelin/img/docs-img/interpreter_binding_mode-scoped.png">
+</div>
+<br/>
+
+In Scoped mode, Zeppelin still runs single interpreter JVM process but multiple Interpreter Group serve each Note.
```
Comment (zjffdu) [inline docs/usage/interpreter/interpreter_binding_mode.md]: >> Zeppelin still runs single interpreter JVM process but multiple Interpreter Group serve each Note. 

This is only correct when set scoped per note. As in another comment, we should mention in which dimension the scope is. And what does it mean for Per User and Per Note. The same for isolated mode.

Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: image

Here is the description for that. If we need to repeat again there, I will add. Do we need?

**Comment by 1ambda** [comment]
@zjffdu Thanks for the review.

Now it looks like

image

image

**Comment by 1ambda** [comment]
CI failed but irrelevant.

**Comment by zjffdu** [comment]
>> So, each Note have their own dedicated session but still it’s possible to share objects between different Interpreter Groups while they’re in the same JVM process.

Is this correct ? AFAIK, only sparkcontext is shared in scoped mode. Variable defined in session 1 can not be used in session 2

>> From the code perspective, there is only one `InterpreterGroup` for the scoped mode, but multiple sessions in one `InterpreterGroup`.

Sorry, I think we can remove this before we find a more proper way to explain. This might confuse users too.

**Comment by 1ambda** [comment]
@zjffdu regarding that, I have question as well.

- In isolated mode / shared mode, can't we share variables via `z.put?` and `z.get`?

**Comment by zjffdu** [comment]
We can, but this is via `ResourcePool`. We might need to say that we could share object via resource pool. Otherwise users might be confused.

**Comment by 1ambda** [comment]
@zjffdu yes, it would be more clear. 

@jongyoul. As zjffdu said, variables (e.g scala variables, ...) in the scope mode can't be shared directly. is it correct?

**Comment by jongyoul** [comment]
Yes, correct. Except `Shared`, user cannot share anything directly.

**Comment by 1ambda** [comment]
@jongyoul Thanks! 

I fixed the incorrect description. Now it looks like

image

**Comment by cacti77** [comment]
> from the code perspective, there's only one InterpreterGroup for scoped mode, but multiple sessions in one InterpreterGroup.

If this is true then surely the diagram for scoped mode should only have one InterpreterGroup in it, but with multiple (Spark?) sessions inside that, with several interpreters inside each session?

Also, this 'per note' and 'per user' does appear to add another dimension. So, to explain what's happening you'd need one table for shared/scoped/isolated for the 'per user' mode and another table again for shared/scoped/isolated for the 'per note' mode.

It would also be good to have a column or table explaining how data can be shared, or not, between notebooks in the 3 different modes. As a Zeppelin user I find all these different combinations really confusing. Maybe an architectural diagram of how the different parts of Zeppelin talk to each other would help here. Or maybe the product itself should be simplified to remove a lot of these combinations, especially if they're not used much?!

**Comment by 1ambda** [comment]
@cacti77 So, you want to add some description for **per user** as well?

**Comment by 1ambda** [comment]
1. Added the **sharing objects** column.
2. Enhanced the description for **per user**

image

3. Added images for UI

image

**Comment by zjffdu** [comment]
@cacti77  @1ambda  How about diagram like this ?

image
image

**Comment by 1ambda** [comment]
@zjffdu I like that diagram, it's more precise and accurate. But would be nice to

- colorize
- and attach note UI so that users easily understand 

like 

image

**Comment by 1ambda** [comment]
@cacti77 @zjffdu Please tell me if this is incorrect

image

**Comment by 1ambda** [comment]
@zjffdu Let me work on the new diagram based on yours, and will share it here.

**Comment by 1ambda** [comment]
Now it looks like

image

image

image

image

image

image

**Comment by felixcheung** [comment]
awesome!

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 2]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
 
 ## Overview
 
+<center><img src="/assets/themes/zeppelin/img/docs-img/interpreter_binding_per_note_user.png" height="70%" width="70%"></center>
+<center><img src="/assets/themes/zeppelin/img/docs-img/interpreter_binding_scoped_isolated.png" height="70%" width="70%"></center>
 
+<br/>
+
+Interpreter is a JVM process that communicates to Zeppelin daemon using thrift. 
+Each Interpreter process can have Interpreter Groups, and each interpreter instance belongs to this Interpreter Group.
```
Comment (felixcheung) [inline docs/usage/interpreter/interpreter_binding_mode.md]: does this mean to be one group?
`Each Interpreter process can have an Interpreter Group`
?
or
`Each Interpreter process can have multiple Interpreter Groups`
for clarity?

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 3]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+In **Shared** mode, single JVM process and single Interpreter Group serves all notes.
+
+## Scoped Mode
+
+<div class="text-center">
+    <img src="/assets/themes/zeppelin/img/docs-img/interpreter_binding_mode-scoped.png">
+</div>
+<br/>
+
+In Scoped mode, Zeppelin still runs single interpreter JVM process but multiple Interpreter Group serve each note. (in case of **per note**)
```
Comment (felixcheung) [inline docs/usage/interpreter/interpreter_binding_mode.md]: `multiple Interpreter Group` -> `multiple Interpreter Groups`

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 4]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+In case of **per user** (available on multi-user environment), Zeppelin manages interpreter sessions per user. For example,
+ 
+- In **scoped + per user** mode, `User A`'s notes will be affected by `User B`'s notes. Because all notes are running on the same JVM
+- On the other hand, **isolated + per user** mode, `User A`'s notes will not be affected by others' notes which running on separated JVMs
+
+<br/>
+
+Each Interpreter implementation may have different characteristics depending on the back end system that they integrate. And 3 interpreter modes can be used differently.
+Let’s take a look how Spark Interpreter implementation uses these 3 interpreter modes with **per note** mote, as an example.
```
Comment (felixcheung) [inline docs/usage/interpreter/interpreter_binding_mode.md]: `mote` -> `note`

**Comment by 1ambda** [comment]
Many sentences should be updated since our major unit is the session, not interpreter group in the screenshots. Let me updated.

**Comment by zjffdu** [comment]
Thanks @1ambda for the help on the updated diagram, it looks create. Just leave a minor comment.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 5]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+<center><img src="/assets/themes/zeppelin/img/docs-img/interpreter_binding_scoped_isolated.png" height="70%" width="70%"></center>
 
+<br/>
+
+Interpreter is a JVM process that communicates to Zeppelin daemon using thrift. 
+Each Interpreter process can have Interpreter Groups, and each interpreter instance belongs to this Interpreter Group.
+(See [here](../../development/writing_zeppelin_interpreter.html) to understand more about its internal structure.) 
+
+Zeppelin provides 3 different modes to run interpreter process: **shared**, **scoped** and **isolated**.   
+Also, user can specify the scope of this mode as well: **per user** or **per note**.
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: The screenshot shows 3 scopes: Globally, Per Note, Per User, so shouldn't Globally be mentioned too? Also, in the UI itself under Option, does it really mean "The interpreter will be instantiated...", or should it be "The interpreter **group** will be instantiated..."?

**Comment by cacti77** [comment]
Sorry, I've never done a PR before so I'm not familiar with how best to contribute my suggestions. I'm also finding it difficult to see what the final docs page would look like as a whole.

1. The diagrams are looking great but shouldn't we also show the InterpreterGroup since it is mentioned in the UI itself and in the text of this help page? Does the InterpreterGroup contain the session, or is it vice versa?

1. In the first set of 3 diagrams maybe distinguish the Interpreters in each session (e.g., Interpreter 1, Interpreter 2, Interpreter 3; don't use A, B, C because you've already used that to distinguish the notes!).

1. Thanks for adding the extra table column: typo (occurs twice): share objets --> share obje**c**ts (and remove ')' after "via ResourcePool")

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 5 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: 1. Yeah. `Globally` means the `shared` mode. 
2. **The interpreter group will be instantiated** more makes sense.

**Comment by 1ambda** [comment]
@cacti77 

Your comment is helpful. You can preview the page like

- [setup rvm 2.1.0+ and install required bundles for docs/](https://github.com/apache/zeppelin/tree/master/docs#build-documentation)
- bundle exec jekyll serve --watch
- open `http://localhost:4000/usage/interpreter/interpreter_binding_mode.html`

1. `1 InterpreterGroup` matchs `1 Interpreter Process`, it might contains multiple sessions (e.g **scope mode**
2. I will update it. 
3. I will udpate it.

**Comment by cacti77** [comment]
@1ambda Thanks for the tip. I was hoping there'd be an easier way to see the web page; oh well:).

> 1 InterpreterGroup matchs 1 Interpreter Process

Are you sure about this?@felixcheung above said:

> Each Interpreter process can have multiple Interpreter Groups

I still think it would be good to distinguish Interpreter Groups from the Interpreter Process in the diagrams. After looking at the code for [InterpreterGroup ](https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterGroup.java) it makes a bit more sense to me now.  One InterpreterGroup can be associated with multiple sessions, and each session id is associated with a list of interpreters. I'm just not sure yet [how the sessions are generated with respect to notes, users, or their combination](https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterGroup.java#L41-L44)

**Comment by 1ambda** [comment]
@cacti77 

1. image

2. Could you provide an example diagram? you can duplicate and edit [this](https://docs.google.com/presentation/d/1Y8lAJFlMO-Jm6ObqcTPmZa9_S9NfyFDsv93xSXvO7l0/edit?usp=sharing) with the description to be included?

**Comment by cacti77** [comment]
@1ambda Thanks; I can try - I've requested edit permission on that doc.

**Comment by cacti77** [comment]
@1ambda Never mind; I'll edit my own copy as you suggested previously.

**Comment by cacti77** [comment]
@1ambda Look [here ](https://docs.google.com/presentation/d/1eIdnReyFxTAAOKTwc_a_mynhlt9mNq6GQAAewFP00PI/edit?usp=sharing) please. I've added 1, 2 or 3 to the Interpreters in the first 3 diagrams. I've only got the patience to add the InterpreterGroup to the very first diagram, but it should be added to all 6 diagrams. In the two scoped-mode diagrams (diagrams 2 and 5) the InterpreterGroup should contain all 3 sessions. In the last 3 diagrams I presume the red SparkContext box should be _outside_ the InterpreterGroup boxes since the code for the latter doesn't appear to directly reference the SparkContext. But @zjffdu should be able to confirm this is true.

Hope this helps you a bit!

**Comment by zjffdu** [comment]
Thanks @cacti77 & @1ambda. The diagram is correct. One interpreter process would have only one InterperterGroup, and InterpreterGroup could have one session for shared/isolated mode, and multiple sessions for scoped mode.

**Comment by 1ambda** [comment]
Thanks for the review.

@felixcheung Resolved all comments.
@zjffdu @cacti77 Update imgs and description. See the last 2 commits.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 6]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+So, each note have absolutely isolated session. (but still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+
+## Which mode should I use?
+
+<br/>
+
+Mode | Each notebook...	| Benefits | Disadvantages | Sharing objects
+--- | --- | --- | --- | ---
+**shared** | Shares a single sessions in a single interpreter process (JVM) | Low resource utilization and Easy to share data between notebooks | All notebooks are affected if Interpreter Process dies | can share directly
+**scoped** | Has its own note sessions in the same Interpreter Process (JVM) | Less resource utilization than isolated mode | All notebooks are affected if Interpreter Process dies | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange))
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Can't share directly, but it's possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 7]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+## Which mode should I use?
+
+<br/>
+
+Mode | Each notebook...	| Benefits | Disadvantages | Sharing objects
+--- | --- | --- | --- | ---
+**shared** | Shares a single sessions in a single interpreter process (JVM) | Low resource utilization and Easy to share data between notebooks | All notebooks are affected if Interpreter Process dies | can share directly
+**scoped** | Has its own note sessions in the same Interpreter Process (JVM) | Less resource utilization than isolated mode | All notebooks are affected if Interpreter Process dies | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+**isolated** | Has its own Interpreter Process | One notebook not affected directly by other notebooks (**per note**) | Can't share data between notebooks easily (**per note**) | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange))
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Can't share directly, but it's possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 8]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
 
 ## Overview
 
+<center><img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_per_note_user.png" height="70%" width="70%"></center>
+<center><img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_scoped_isolated.png" height="70%" width="70%"></center>
 
+<br/>
+
+Interpreter is a JVM process that communicates with Zeppelin daemon using thrift. 
+Each Interpreter process can have a interpreter group, and each interpreter instance belongs to this interpreter group.
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Each interpreter process has a single interpreter group, and this interpreter group can have one or more instances of an interpreter.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 9]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+(See [here](../../development/writing_zeppelin_interpreter.html) to understand more about its internal structure.) 
+
+Zeppelin provides 3 different modes to run interpreter process: **shared**, **scoped** and **isolated**.   
+Also, user can specify the scope of these mode as well: **per user** or **per note**.
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Also, the user can specify the scope of these modes as well: **per user** or **per note**.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 10]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+Interpreter is a JVM process that communicates with Zeppelin daemon using thrift. 
+Each Interpreter process can have a interpreter group, and each interpreter instance belongs to this interpreter group.
+(See [here](../../development/writing_zeppelin_interpreter.html) to understand more about its internal structure.) 
+
+Zeppelin provides 3 different modes to run interpreter process: **shared**, **scoped** and **isolated**.   
+Also, user can specify the scope of these mode as well: **per user** or **per note**.  
+These 3 modes give flexibility to fit Zeppelin into any type of use cases.
+
+In this documentation, we mainly discuss the combination of **per note** mode with **shared**, **scoped** and **isolated** modes for explanation.
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: In this documentation, we mainly discuss the **per note** scope in combination with the **shared**, **scoped** and **isolated** modes.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 11]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+## Scoped Mode
+
+<div class="text-center">
+    <img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_mode-scoped.png">
+</div>
+<br/>
+
+In Scoped mode, Zeppelin still runs single interpreter JVM process but multiple sessions serve each note. (in case of **per note**) 
+So, each note have their own dedicated session. (but still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange))
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: In Scoped mode, Zeppelin still runs a single interpreter JVM process but, in the case of **per note** scope, each note runs in its own dedicated session. (Note it is still possible to share objects between these notes via [ResourcePool](../../interpreter/spark.html#object-exchange).)

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 12]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+## Isolated Mode
+
+<div class="text-center">
+    <img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_mode-isolated.png">
+</div>
+<br/>
+
+**Isolated** mode runs separate interpreter process for each note. (in case of **per note**) 
+So, each note have absolutely isolated session. (but still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange))
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: **Isolated** mode runs a separate interpreter process for each note in the case of **per note** scope.
So, each note has an absolutely isolated session. (But it is still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange).)

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 13]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+**Isolated** mode runs separate interpreter process for each note. (in case of **per note**) 
+So, each note have absolutely isolated session. (but still possible to share objects via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+
+## Which mode should I use?
+
+<br/>
+
+Mode | Each notebook...	| Benefits | Disadvantages | Sharing objects
+--- | --- | --- | --- | ---
+**shared** | Shares a single sessions in a single interpreter process (JVM) | Low resource utilization and Easy to share data between notebooks | All notebooks are affected if Interpreter Process dies | can share directly
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: +**shared** | Shares a single session in a single interpreter process (JVM) | Low resource utilization and it's easy to share data between notebooks | All notebooks are affected if the interpreter process dies | Can share directly

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 14]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+In case of **per user** (available on multi-user environment), Zeppelin manages interpreter sessions per user. For example,
+ 
+- In **scoped + per user** mode, `User A`'s notes **might** be affected by `User B`'s notes. (e.g JVM dies, ...) Because all notes are running on the same JVM
+- On the other hand, **isolated + per user** mode, `User A`'s notes will not be affected by others' notes which running on separated JVMs
+
+<br/>
+
+Each Interpreter implementation may have different characteristics depending on the back end system that they integrate. And 3 interpreter modes can be used differently.
+Let’s take a look how Spark Interpreter implementation uses these 3 interpreter modes with **per note** mdoe, as an example.
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Let’s take a look how Spark Interpreter implementation uses these 3 interpreter modes with **per note** mode, as an example.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 15]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+In this mode, if `Note A` defines variable ‘a’ then `Note B` not only able to read variable ‘a’ but also able to override the variable.
+
+<div class="text-center">
+    <img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_mode-example-spark-scoped.png">
+</div>
+<br/>
+
+In Scoped mode, each note has its own Scala REPL. 
+So variable defined in a note can not be read or overridden in another note. 
+However, still single SparkContext serves all the sessions.
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: However, a single SparkContext still serves all the sessions.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 16]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+<div class="text-center">
+    <img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_mode-example-spark-scoped.png">
+</div>
+<br/>
+
+In Scoped mode, each note has its own Scala REPL. 
+So variable defined in a note can not be read or overridden in another note. 
+However, still single SparkContext serves all the sessions. 
+And all the jobs are submitted to this SparkContext and fair scheduler schedules the job.
```
Comment (cacti77) [inline docs/usage/interpreter/interpreter_binding_mode.md]: And all the jobs are submitted to this SparkContext and the fair scheduler schedules the jobs.

**Comment by cacti77** [comment]
@1ambda Thanks for finishing the diagrams. I would probably put the SparkContext inside the interpreter process but outside the InterpreterGroup; but if @zjffdu is happy with the diagrams I guess it doesn't matter. I've gone through interpreter_binding_mode.md, tidying up the English some more. I'm done with my edits now, thanks.

**Comment by cacti77** [comment]
Apologies @1ambda for not interacting with this PR properly; I may have overlooked some of your recent edits in the version of interpreter_binding_mode.md I commented on yesterday.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 6 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 7 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 8 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 9 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 10 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 11 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 12 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 13 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 17]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
...
+
+<br/>
+
+Mode | Each notebook...	| Benefits | Disadvantages | Sharing objects
+--- | --- | --- | --- | ---
+**shared** | Shares a single sessions in a single interpreter process (JVM) | Low resource utilization and Easy to share data between notebooks | All notebooks are affected if Interpreter Process dies | can share directly
+**scoped** | Has its own note sessions in the same Interpreter Process (JVM) | Less resource utilization than isolated mode | All notebooks are affected if Interpreter Process dies | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+**isolated** | Has its own Interpreter Process | One notebook not affected directly by other notebooks (**per note**) | Can't share data between notebooks easily (**per note**) | can't share directly, but possible to share objets via [ResourcePool](../../interpreter/spark.html#object-exchange)) 
+
+In case of **per user** (available on multi-user environment), Zeppelin manages interpreter sessions per user. For example,
```
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 14 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 15 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

File: docs/usage/interpreter/interpreter_binding_mode.md (hunk 16 above)
Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Fixed.

**Comment by 1ambda** [comment]
resolved @cacti77's all comments

File: docs/usage/interpreter/interpreter_binding_mode.md [hunk 18]
```diff
@@ -25,4 +25,97 @@ limitations under the License.
 
 ## Overview
 
+<center><img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_per_note_user.png" height="70%" width="70%"></center>
+<center><img src="{{BASE_PATH}}/assets/themes/zeppelin/img/docs-img/interpreter_binding_scoped_isolated.png" height="70%" width="70%"></center>
 
+<br/>
+
+Interpreter is a JVM process that communicates with Zeppelin daemon using thrift.
```
Comment (zjffdu) [inline docs/usage/interpreter/interpreter_binding_mode.md]: Interpreter -> Interpreter Process

Comment (1ambda) [inline docs/usage/interpreter/interpreter_binding_mode.md]: fixed.

**Comment by 1ambda** [comment]
resolved @zjffdu's comment

**Comment by 1ambda** [comment]
Please let me know if we need to add / improve / modify something.

**Comment by zjffdu** [comment]
LGTM

**Comment by cacti77** [comment]
@1ambda And it looks good to me too; thanks for being patient and making all those changes!

**Comment by 1ambda** [comment]
it was an honor to make more correct, concise documentation with your help. 

Merge if no more discussion.

**Comment by cacti77** [comment]
@1ambda One question please: Do these docs constitute a new web page in Zeppelin's docs? I'm just wondering how a user would navigate to this new information on binding modes from the Docs home page. Will the new page be linked to from under http://zeppelin.apache.org/docs/0.7.2/manual/interpreters.html#interpreter-binding-mode for example? Or will it become part of the interpreters.html page itself?

**Comment by 1ambda** [comment]
@cacti77 it's provided as of 0.8.0-SNAPSHOT. 

- http://zeppelin.apache.org/docs/0.8.0-SNAPSHOT/usage/interpreter/interpreter_binding_mode.html

here are the background

- we re-organized contents structure of docs in 0.8.0-SNAPSHOT
- the interpreter binding page was created at that time (so it's currently empty)

**Comment by cacti77** [comment]
Thanks @1ambda . So, looking at #2371, it looks like a user would simply click on the dropdown and then Interpreter > Interpreter Binding Mode to access the new web page. Is that correct? Basically, as long as all the info about interpreter binding modes is in one place and easy to find, that's all I care about:).

**Comment by 1ambda** [comment]
Sure, users can easily access the doc.

**Comment by 1ambda** [comment]
Deployed 

- http://zeppelin.apache.org/docs/0.8.0-SNAPSHOT/usage/interpreter/interpreter_binding_mode.html

**Pull request closed without merging**